### DISCUSSION TITLE
How to trigger OnRecordBeforeCreateRequest when creating record inside another hook

### DISCUSSION THREAD

**Discussion by thangngoc89** [pocketbase/pocketbase#4868] [Q&A] [answered]
I'm extending Pocketbase using Golang and I have a central `queues` collection for queueing async tasks with read-only API rules for user. 

Let's say that user could indirectly trigger a queue for image export by writing to `image_exports` collection.
The `image_exports` collection has a before create request hook that create new row to the `queues` collection like below:

```go
app.OnRecordBeforeCreateRequest("image_exports").Add(func(e *core.RecordCreateEvent) error {
		exportId := e.Record.Id
		if exportId == "" {
			exportId = security.RandomString(15)
		}
		queueId := security.RandomString(15)
		e.Record.Set("id", exportId)
		e.Record.Set("queue", queueId)

		// Create new row in queue collection
		queueCollection, err := app.Dao().FindCollectionByNameOrId("queues")
		if err != nil {
			return err
		}

		record := models.NewRecord(queueCollection)

		form := forms.NewRecordUpsert(app, record)

		form.LoadData(map[string]any{
			"id":         queueId,
			"status":     "NEW",
			"model_name": "image_exports",
			"model_id":   exportId,
		})

		if err := form.Submit(); err != nil {
			return err
		}

		app.Logger().Debug("Emit queue", "created new queue entry", record.Id)
		return nil
	})
```

After that I have another hook for the `queues` collection to fire off the enqueue logic

```go
	app.OnRecordBeforeCreateRequest("queues").Add(func(e *core.RecordCreateEvent) error {
		app.Logger().Debug("Queue hook", "hook started", e.Record.Id)
	        // actual enqueue logic
		return nil
	})
```

The `queues` hook isn't executed when the record was created inside the `image_exports` hook. It's executed only when creating a new `queues` directly using the API.

Is there anyway I could manually trigger the hook for `queues` collection?

**Comment by ganigeorgiev** [comment, marked as answer]
> Is there anyway I could manually trigger the hook for queues collection?

`*Request` hooks fires only on requests to the specific endpoint.  You can trigger it manually using the `Trigger` method but you'll have to construct somehow its expected event object (including the `echo.Context` available as `e.HttpContext`).

If your code doesn't rely on the request data an alternative could be to:
- Export your functions as separate methods and call them manually.
- Attach to the [`OnModel*` hooks](https://pocketbase.io/docs/go-event-hooks/#db-hooks) instead (they fire on each db model change).
  Note that the `OnModel*` hooks operates with the `Model` interface and you'll need type assertion (ex. `record := e.Model.(*models.Record)`). With the [refactoring](https://github.com/pocketbase/pocketbase/discussions/4355) the type assertion will no longer be necessary as there are dedicated `OnRecord*` db hooks.

**Reply by thangngoc89** [reply]
Thank you for your super helpful answer. I think the first option which is calling the enqueue logic directly in the `image_exports` hook would be best.
